### ISSUE TITLE
LanguageInvoker segfault

### ISSUE THREAD

**Issue by mrpg99** [xbmc/xbmc#19961]
Tried my luck on the forum first, but after 80 views, no reply.

I am currently on these versions :
Ubuntu 21.04
Kernel 5.11.0-22-generic
Kodi 20.0-ALPHA1 (19.90.101) Git:20210709-cad98e3da3

My kodi or ubuntu has started crashing, i am not sure what has caused this, i did have some unsupported plugins installed, which i have now tried to remove.

But its still crashing, and it does not matter if i go back to kodi 19.1 or try the latest nigthly version.

I get this in /var/log/syslog :

Jul 10 12:30:15 mediapc kernel: [  177.645872] LanguageInvoker[3718]: segfault at 0 ip 00007fae4053df45 sp 00007fadf17f81e8 error 4 in libc-2.33.so[7fae403e4000+16b000]
Jul 10 12:30:15 mediapc kernel: [  177.645883] Code: 00 00 0f 1f 00 31 c0 c5 f8 77 c3 66 2e 0f 1f 84 00 00 00 00 00 f3 0f 1e fa 89 f9 48 89 fa c5 f9 ef c0 83 e1 3f 83 f9 20 77 2b <c5> fd 74 0f c5 fd d7 c1 85 c0 0f 85 eb 00 00 00 48 83 c7 20 83 e1
Jul 10 12:32:24 mediapc systemd[1]: Starting Download data for packages that failed at package install time...
Jul 10 12:32:24 mediapc systemd[1]: update-notifier-download.service: Succeeded.
Jul 10 12:32:24 mediapc systemd[1]: Finished Download data for packages that failed at package install time.

And i get these crashlogs in my home directory :

https://pastebin.com/jesuTAS8

Here is my logile :

https://pastebin.com/Nv6sev1n

Any help is very appriciated

Edit, forgot to mention, that i searched for this segfault error, and found suggestions that it could be the gpu drivers, so i have also enable this repo, and installed the intel drivers, but its still crashing :
https://launchpad.net/~oibaf/+archive/ub...cs-drivers

**Comment by lrusak** [comment]
You need an actual stack trace which involves installing the debug symbols and gdb. The current crash log doesn't actually show anything.

**Comment by mrpg99** [comment]
can you help with instructions for ubuntu 21.04, i.e how i install the debug symbols and gdb?
i tried following this guide : https://www.hiroom2.com/2019/05/22/ubunt...dbgsym-en/

but i am still not getting a stack trace in the crashlogs

i have also since the first post, removed every single plugin, but its still crashing.

Br
Patric

**Comment by mrpg99** [comment]
How can we progress this one? I desperately need help!

**Comment by basilgello** [comment]
@wsnipex arent there -dbgsym packages in PPA?

@mrpg99 can you run `gdb /usr/lib/x86_64-linux-gnu/kodi/kodi.bin` and reprodice an error? When GDB stops on segmentation fault, do `bt full` and post it here.

**Comment by mrpg99** [comment]
Hi,
Yesterday i got the advice to move my .kodi directory, and start from scratch.
So i did this, as soon as i moved my .kodi dir, it stopped crashing.

I then changed/added this to kodi 

1 installed aeon nox silvo, and customized the skin and kodi settings
2 installed youtube,svtplay,retrospect,formula1,DR TV,
3 the only two plugins i installed that were not available in the default kodi repo were Netflix & Discovery Plus
4 installed pvr simple client and added my channels

the system was not crashing after all of the above, we watched a movie, and after this went onto youtube to search for some videos, and now it started crashing again.

I now get a stacktrace in the crashlog.

can you please have a look at this https://paste.ubuntu.com/p/Rn628JZnSP/

Best Regards
Patric

**Comment by basilgello** [comment]
Better!

Can you reproduce once more but instead of `bt full` do `py-bt-full` ?

**Comment by mrpg99** [comment]
Hi,
in my last post i just posted the crashlog that had stacktra in it.

when i try to run it with gdb 

https://pastebin.ubuntu.com/p/6c5jBpDFRB/

seems i dont have the needed symbols, what can i do to correct this?
also kodi does not crash if i start it from my mate desktop (i usually autostart kodi using lightdm)

Br
Patric

**Comment by basilgello** [comment]
TBH I expected debug symbols to be present in PPA as separate package. Then let me build a package for you I guess :) Also I think I will raise the question about -dbg / -dbgsym package to be included in PPAs.

**Comment by mrpg99** [comment]
thanks for your reply basilgello!
I am very grateful if you can help with the debug symbols etc!

Br
Patric

**Comment by basilgello** [comment]
@mrpg99 lets communucate here :) I just noticed you tried 20.0 alpha :) What if I build **my** nightly targeting hirsute 21.04 for you rather tgan 19.1?

**Comment by wsnipex** [comment]
did you already try with a clean userdata dir(~/.kodi)?

**Comment by basilgello** [comment]
https://fex.net/s/tx1xxtm @mrpg99

**Comment by mrpg99** [comment]
> did you already try with a clean userdata dir(~/.kodi)?

Yes, please see my post above, works for a while then starts crashing again.

**Comment by mrpg99** [comment]
> https://fex.net/s/tx1xxtm @mrpg99

Thanks, could you please let me know what this is? And what I should try?

Br

Patric

**Comment by mrpg99** [comment]
Hi,
Sorry for slow response,

i have now downloaded your package, but i tun into problems 

mrpg@mediapc:~/amd64$ sudo dpkg -i kodi_*.deb kodi-bin_*.deb kodi-data*.deb kodi-repository-kodi_*.deb kodi_*.deb
(Reading database ... 294645 files and directories currently installed.)
Preparing to unpack kodi_20.0~alpha0+git20210711.1272608+dfsg1-1_amd64.deb ...
Unpacking kodi (2:20.0~alpha0+git20210711.1272608+dfsg1-1) over (2:19.1+git20210509.0555-85e05228b4-0~hirsute) ...
dpkg: error processing archive kodi_20.0~alpha0+git20210711.1272608+dfsg1-1_amd64.deb (--install):
 trying to overwrite '/usr/bin/kodi', which is also in package kodi-bin 2:19.1+git20210509.0555-85e05228b4-0~hirsute
Preparing to unpack kodi-bin_20.0~alpha0+git20210711.1272608+dfsg1-1_amd64.deb ...
Unpacking kodi-bin (2:20.0~alpha0+git20210711.1272608+dfsg1-1) over (2:19.1+git20210509.0555-85e05228b4-0~hirsute) ...
Selecting previously unselected package kodi-data.
Preparing to unpack kodi-data_20.0~alpha0+git20210711.1272608+dfsg1-1_all.deb ...
Unpacking kodi-data (2:20.0~alpha0+git20210711.1272608+dfsg1-1) ...
Replaced by files in installed package kodi (2:19.1+git20210509.0555-85e05228b4-0~hirsute) ...
Selecting previously unselected package kodi-repository-kodi.
Preparing to unpack kodi-repository-kodi_20.0~alpha0+git20210711.1272608+dfsg1-1_all.deb ...
Unpacking kodi-repository-kodi (2:20.0~alpha0+git20210711.1272608+dfsg1-1) ...
Replaced by files in installed package kodi (2:19.1+git20210509.0555-85e05228b4-0~hirsute) ...
Preparing to unpack kodi_20.0~alpha0+git20210711.1272608+dfsg1-1_amd64.deb ...
Unpacking kodi (2:20.0~alpha0+git20210711.1272608+dfsg1-1) over (2:19.1+git20210509.0555-85e05228b4-0~hirsute) ...
dpkg: dependency problems prevent configuration of kodi-bin:
 kodi-bin depends on libgif7 (>= 5.1); however:
  Package libgif7 is not installed.
 kodi-bin depends on libsndio7.0 (>= 1.1.0); however:
  Package libsndio7.0 is not installed.

dpkg: error processing package kodi-bin (--install):
 dependency problems - leaving unconfigured
dpkg: dependency problems prevent configuration of kodi-data:
 kodi-data depends on libjs-iscroll; however:
  Package libjs-iscroll is not installed.

dpkg: error processing package kodi-data (--install):
 dependency problems - leaving unconfigured
Setting up kodi-repository-kodi (2:20.0~alpha0+git20210711.1272608+dfsg1-1) ...
dpkg: dependency problems prevent configuration of kodi:
 kodi depends on kodi-bin (>= 2:20.0~alpha0+git20210711.1272608+dfsg1-1); however:
  Package kodi-bin is not configured yet.
 kodi depends on kodi-bin (<< 2:20.0~alpha0+git20210711.1272608+dfsg1-1.1~); however:
  Package kodi-bin is not configured yet.
 kodi depends on kodi-data (>= 2:20.0~alpha0+git20210711.1272608+dfsg1-1); however:
  Package kodi-data is not configured yet.

dpkg: error processing package kodi (--install):
 dependency problems - leaving unconfigured
Processing triggers for man-db (2.9.4-2) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
Processing triggers for mailcap (3.68ubuntu1) ...
Processing triggers for gnome-menus (3.36.0-1ubuntu1) ...
Processing triggers for desktop-file-utils (0.26-1ubuntu1) ...
Processing triggers for bamfdaemon (0.5.5+21.04.20210119-0ubuntu1) ...
Rebuilding /usr/share/applications/bamf-2.index...
Processing triggers for menu (2.1.47ubuntu4) ...
Errors were encountered while processing:
 kodi_20.0~alpha0+git20210711.1272608+dfsg1-1_amd64.deb
 kodi-bin
 kodi-data
 kodi

**Comment by basilgello** [comment]
@mrpg99 I noticed you had issues installing package. Did you solve the issue by removing 19.1 and installing 20.0 cleanly?

Also, you ran full debugging session, but Kodi failed to create GUI this time. If you solved package installation errors, but still have `failed to create GUI` error, make sure you have `DISPLAY` environment set and pointing to your X server. Or do you use Wayland ?

**Comment by mrpg99** [comment]
Hi, again sorry for my slow response.

i managed to fix the dependancies, i did not remove kodi 19 first, just installed your packages ontop.

last time i ran the command through ssh, and forgot to forward x

i use the kodi-standalone service to start kodi on boot : https://github.com/graysky2/kodi-standalone-service

to run kodi under gdb like you suggested, i enabled lightdm again, and now manually started kodi under gdb

it does not crash in this mode... 

when i enable the kodi standalone service, and rebooed, it has not crashed yet, so i will get back tou you later today if it starts crashing again.

Br
Patric

**Comment by cvandesande** [comment]
Not the original reporter, but do suffer from this issue. For some reason it's really hard to reproduce this issue under gdb. I've been trying all week. Outside of gdb it happens at random times, a few times per hour.  Finally got the error today. Here is the log (with a bit of info redacted). I really hope this is helpful
[gdb.log](https://github.com/xbmc/xbmc/files/7679186/gdb.log)

**Comment by elgatito** [comment]
In my case gdb shows different backtrace:
https://pastebin.com/3DWnJx0R

**Comment by elgatito** [comment]
In my case I have installed python-dbg and viewed with py-bt, that shows this:

```
(gdb) py-bt
Traceback (most recent call first):
  File "/usr/lib/python3.9/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
  File "/usr/lib/python3.9/ssl.py", line 1552, in _create
  File "/usr/lib/python3.9/ssl.py", line 756, in wrap_socket
    return context
  File "/usr/lib/python3.9/http/client.py", line 1682, in connect
  File "/usr/lib/python3.9/http/client.py", line 952, in send
    self.connect()
  File "/usr/lib/python3.9/http/client.py", line 1012, in _send_output
    self.send(msg)
  File "/usr/lib/python3.9/http/client.py", line 1252, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1303, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1257, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.9/urllib/request.py", line 1602, in do_open
    
  File "/usr/lib/python3.9/urllib/request.py", line 1645, in https_open
    #
  File "/usr/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 1302, in _open
    
  File "/usr/lib/python3.9/urllib/request.py", line 517, in open
    response = self._open(req, data)
  File "/usr/lib/python3.9/urllib/request.py", line 1771, in http_error_302
    urltype, proxyhost = _splittype(proxy)
  File "/usr/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 555, in error
    result = self._call_chain(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 888, in http_response
    if base[0] != test[0]:
  File "/usr/lib/python3.9/urllib/request.py", line 523, in open
    response = meth(req, response)
  File "/usr/lib/python3.9/urllib/request.py", line 982, in urlopen
    yield (scheme, None)
  File "/home/elgato/.kodi/addons/plugin.video.tam/proxy_list.py", line 664, in TEST2PRX
    response = urllib2.urlopen(req, timeout=1)
  File "/home/elgato/.kodi/addons/plugin.video.tam/proxy_list.py", line 644, in run
    if TEST2PRX('http://ya.ru', self.param['prx']): update_Ltest(self.param['prx'])
  File "/usr/lib/python3.9/threading.py", line 954, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.9/threading.py", line 912, in _bootstrap
    self._bootstrap_inner()
(gdb)
```

Am I understanding that right, that segfault is caused by Requests python module?

**Comment by fuzzard** [comment]
The bulk of this issue is due to how python and it's modules handle sub interpreters.
There are several open python bpos and have been many prs to cpython to handle sub interpreters better, but it's still very much a work in progress.

All C modules are/were being updated to what they call multi stage init, which essentially moves away from static init of members. Sub interpreters can modify these static members which can cause other sub interpreter states to crash horribly.

For the user with the SSL crash running python 3.9, if you go to 3.10, the SSL module was converted to multistage init, and shouldn't crash in the same manner for that particular module. However now all modules are converted, so there are still failure points on other modules.

This is a cpython issue, and there's not much we can do but wait for it to be resolved in cpython

**Comment by elgatito** [comment]
@fuzzard Totally agree that this isn't Kodi issue at the end.

It would be very useful to have something like `py-bt` in the crash log, since it is possible to crash whole Kodi with a simple python script.

**Comment by garbear** [comment]
This should be fixed in the latest Kodi v21 with Python 3.12.

**Comment by basilgello** [comment]
And in 20.5 in Debian sid/trixie
